File: backend/src/main/java/com/fincoach/service/JwtService.java
package com.fincoach.service;

import com.fincoach.model.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.util.Date;

/**
 * Manages our application's JWT tokens (NOT the provider's tokens).
 *
 * After a successful OAuth2 login, we generate our own JWT and send it to
 * the Angular frontend. The frontend then sends it back in every API request
 * via the "Authorization: Bearer {token}" header.
 *
 * The JWT contains:
 *  - sub    : our internal userId (UUID)
 *  - email  : user's email
 *  - name   : user's display name
 *  - picture: avatar URL (optional)
 *  - iat    : issued at timestamp
 *  - exp    : expiration timestamp (24h by default)
 */
@Service
public class JwtService {

    private static final Logger log = LoggerFactory.getLogger(JwtService.class);

    // Read from application.properties — must be a Base64-encoded 256-bit key
    @Value("${jwt.secret}")
    private String secret;

    // Expiration in milliseconds (default: 86400000 = 24 hours)
    @Value("${jwt.expiration:86400000}")
    private long expiration;

    /**
     * Derives the HMAC-SHA256 signing key from the Base64-encoded secret.
     * Keys.hmacShaKeyFor() validates that the key is long enough for HS256.
     */
    private SecretKey getSigningKey() {
        return Keys.hmacShaKeyFor(Decoders.BASE64.decode(secret));
    }

    /**
     * Generates a signed JWT for the given user.
     * The token is valid for `jwt.expiration` milliseconds.
     */
    public String generateToken(User user) {
        Date now = new Date();
        Date expiresAt = new Date(now.getTime() + expiration);

        String token = Jwts.builder()
                .subject(user.getId())
                .claim("email", user.getEmail())
                .claim("name", user.getName())
                .claim("picture", user.getPictureUrl())
                .issuedAt(now)
                .expiration(expiresAt)
                .signWith(getSigningKey())
                .compact();

        log.debug("JWT generated for userId={}, expires={}", user.getId(), expiresAt);
        return token;
    }

    /**
     * Returns true if the token has a valid signature and is not expired.
     */
    public boolean isValid(String token) {
        try {
            getClaims(token);
            return true;
        } catch (Exception e) {
            log.debug("JWT validation failed: {}", e.getMessage());
            return false;
        }
    }

    /**
     * Parses and returns the claims payload from a token.
     * Throws an exception if the token is invalid or expired.
     */
    public Claims getClaims(String token) {
        return Jwts.parser()
                .verifyWith(getSigningKey())
                .build()
                .parseSignedClaims(token)
                .getPayload();
    }

    /** Extracts the userId (sub) from the token. */
    public String getUserId(String token) {
        return getClaims(token).getSubject();
    }
}

File: backend/src/services/jwt.service.ts
import jwt from 'jsonwebtoken';
import { JwtClaims } from '../types';
import { CONSTANTS } from '../config/constants';
import { createLogger } from '../utils/logger';

const logger = createLogger('JwtService');

/**
 * JWT service — exact port of Java JwtService.java.
 *
 * Generates and verifies our application JWT tokens (NOT the OAuth2 provider's tokens).
 * Token lifetime: 24 hours.
 *
 * The JWT secret must be a Base64-encoded 256-bit key.
 * Generate one with: openssl rand -base64 32
 */
class JwtService {
  private get secret(): Buffer {
    const s = process.env.JWT_SECRET;
    if (!s) {
      logger.error('JWT_SECRET environment variable is not set — token operations will fail');
      throw new Error('JWT_SECRET environment variable is not set.');
    }
    return Buffer.from(s, 'base64');
  }

  /**
   * Generates a signed JWT for the given user.
   */
  generateToken(user: {
    id: string;
    email: string;
    name?: string | null;
    pictureUrl?: string | null;
    role?: string | null;
    firstName?: string | null;
    lastName?: string | null;
    age?: number | null;
    emailVerified?: boolean;
  }): string {
    const payload: Partial<JwtClaims> = {
      email: user.email,
      name: user.name ?? `${user.firstName ?? ''} ${user.lastName ?? ''}`.trim(),
      picture: user.pictureUrl ?? undefined,
      role: user.role ?? 'USER',
      firstName: user.firstName ?? undefined,
      lastName: user.lastName ?? undefined,
      age: user.age ?? undefined,
      emailVerified: user.emailVerified ?? false,
    };

    const token = jwt.sign(payload, this.secret, {
      subject: user.id,
      expiresIn: CONSTANTS.JWT_EXPIRY_MS / 1000, // jwt library expects seconds
      algorithm: 'HS256',
    });

    const expiresAt = new Date(Date.now() + CONSTANTS.JWT_EXPIRY_MS).toISOString();
    logger.debug('JWT generated', { userId: user.id, email: user.email, expiresAt });

    return token;
  }

  /**
   * Verifies and decodes a JWT. Throws if invalid or expired.
   */
  verify(token: string): JwtClaims {
    try {
      const claims = jwt.verify(token, this.secret) as JwtClaims;
      logger.debug('JWT verified', { userId: claims.sub, email: claims.email });
      return claims;
    } catch (err) {
      const jwtErr = err as Error;
      logger.warn('JWT verification failed', {
        reason: jwtErr.name,
        message: jwtErr.message,
      });
      throw err;
    }
  }

  isValid(token: string): boolean {
    try {
      this.verify(token);
      return true;
    } catch {
      return false;
    }
  }
}

export const jwtService = new JwtService();
